glfs.c

```diff
@@ -626,17 +627,58 @@ static int tcmu_glfs_write(struct tcmu_device *dev,
 static int tcmu_glfs_reconfig(struct tcmu_device *dev,
                               struct tcmulib_cfg_info *cfg)
 {
+	struct glfs_state *gfsp;
+	char *config;
+	int ret = -1;
+
 	switch (cfg->type) {
 	case TCMULIB_CFG_DEV_SIZE:
-		/*
-		 * Let glusterfs tools handle size checks
-		 */
-		return 0;
+
+		gfsp = calloc(1, sizeof(*gfsp));
+		if (!gfsp) {
+			goto notify;
+		}
+
+		config = tcmu_get_path(dev);
+		if (!config) {
+			tcmu_dev_err(dev, "tcmu_get_path() failed\n");
+			goto fail;
+		}
+
+		gfsp->fs = tcmu_create_glfs_object(dev, config, &gfsp->hosts);
+		if (!gfsp->fs) {
+			tcmu_dev_err(dev,
+			             "tcmu_create_glfs_object(config=%s) failed\n", config);
+			goto fail;
+		}
+
+		gfsp->gfd = glfs_open(gfsp->fs, gfsp->hosts->path, ALLOWED_BSOFLAGS);
+		if (!gfsp->gfd) {
+			tcmu_dev_err(dev, "glfs_open(vol=%s, file=%s) failed: %m\n",
+			             gfsp->hosts->volname, gfsp->hosts->path);
+			goto out;
+		}
+
+		ret = 0;
+		break;
 	case TCMULIB_CFG_DEV_CFGSTR:
 	case TCMULIB_CFG_WRITE_CACHE:
 	default:
 		return -EOPNOTSUPP;
 	}
+
+out:
+	gluster_cache_refresh(gfsp->fs, tcmu_get_path(dev));
+	gluster_free_server(&gfsp->hosts);
+fail:
+	free(gfsp);
+
+notify:
+	if (ret) {
+		tcmu_notify_conn_lost(dev);
```

Comment:
This function is supposed to do the right thing for the handler. It will wait for running IO, then call handler->close(), and then call handler->open until that succeeds. Above if we have done free(gfsp) already then I think that could do a double free when tcmu_glfs_close is called.

In that email thread when I made the comment about doing it from this context instead of adding the READ CAP callout I thought you just needed to do a glfs_lstat, so I thought you could just do
'''

if (glfs_lstat() == some net error like ENOTCOONN)
       tcmu_notify_conn_lost()
else
       if (st.st_size == cfg->data.dev_size)
                return ok
      else
               return mismatch

'''

If you need to reopen the device to resize it then you need to do block, then wait for aio and then flush (basically take the top of tcmu_acquire_dev_lock and stick in in a new function an call it).